File: 0698-partition-to-k-equal-sum-subsets/0698-partition-to-k-equal-sum-subsets.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var canPartitionKSubsets = function(nums, k) {
    const n = nums.length
    const sum = nums.reduce((acc, val) => acc + val, 0)
    const kSum = sum / k
    if (sum % k !== 0) return false
    const dp = {}
    return helper(Array(n).fill(false), k, kSum)
    
    function helper(used, k, curSum) {
        if ([used, k, curSum] in dp) return dp[[used, k, curSum]]
        if (k === 0) {
            return used.every(u => u === true)
        }
        if (curSum === 0) {
            return helper(used, k-1, kSum)
        }
        for (let i=0; i < n; i++) {
            if (used[i] === false && nums[i] <= curSum) {
                used[i] = true
                if (helper(used, k, curSum - nums[i])) {
                    return dp[[used, k, curSum]] = true
                }
                used[i] = false
            }
        }
        return dp[[used, k, curSum]] = false
    }
};

File: 0698-partition-to-k-equal-sum-subsets/0698-partition-to-k-equal-sum-subsets.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        
        n = len(nums)
        total_sum = sum(nums)
        target = total_sum / k
        dp = {}
        
        if total_sum % k != 0:
            return False
        
        def helper(used: List[int], k: int, cur_sum: int) -> bool:
            
            key = f"""{" ".join(str(s) for s in used)} {k} {cur_sum}"""
            
            if key in dp:
                return dp[key]
            
            if k == 0:
                return sum(used) == n
        
            if cur_sum == 0:
                return helper(used, k-1, target)
            
            for i in range(n):
                if used[i] == 0 and nums[i] <= cur_sum:
                    used[i] = 1
                    if helper(used, k, cur_sum - nums[i]):
                        return True
                    used[i] = 0
            
            dp[key] = False
            return dp[key]
        
        return helper([0] * n, k, target)
            
            